File: 2023/day1.js
import { open } from 'node:fs/promises';

let ln = 0;
try {
	let file = await open("input/day1");
	let content = await file.readFile({ "encoding": "utf-8" });
	file.close();

	let lines = content.split('\n');

	let sum = 0;
	for (const line of lines) {

		if (line.length == 0) {
			continue;
		}

		let tens = undefined;
		let ones = undefined;

		for (let i = 0; i < line.length; ++i) {
			let charCode = line.charCodeAt(i);
			let digited = charCode - 0x30;

			if (digited >= 0 && digited < 10) {
				if (!tens) {
					tens = digited;
				} else {
					ones = digited;
				}
			}
		}

		if (!tens) {
			throw new Error("tens not set!")
		} else {
			if (!ones) {
				ones = tens;
			}

			let num = tens * 10 + ones;
			sum += num;
		}

		ln++;
	}

	console.log(sum);
} catch (err) {
	console.log(`[ln:${ln}]: ${err}`);
}

File: 2023/src/bin/day1.rs
use std::str::FromStr;

use aoc2023::{ParseFile, RuntimeError};

pub struct Calibration {
	pub lines: Vec<String>,
}

impl FromStr for Calibration {
	type Err = RuntimeError;

	fn from_str(s: &str) -> Result<Self, Self::Err> {
		Ok(Self {
			lines: s.lines().map(|s| s.to_owned()).collect(),
		})
	}
}

fn main() -> Result<(), RuntimeError> {
	let cal = Calibration::parse_file("input/day1")?;

	let mut sum = 0;
	for (ln, line) in cal.lines.iter().enumerate() {
		let (mut tens, mut ones) = (None, None);

		for char in line.chars() {
			if let Some(digit) = char.to_digit(10) {
				if tens.is_some() {
					ones = Some(digit);
				} else {
					tens = Some(digit);
				}
			}
		}

		match (tens, ones) {
			(Some(tens), None) => {
				sum += tens * 10 + tens;
			}
			(Some(tens), Some(ones)) => {
				sum += tens * 10 + ones;
			}
			(None, _) => {
				eprintln!("{ln}: didn't have a digit!");
				std::process::exit(1);
			}
		}
	}

	println!("sum = {sum}");

	Ok(())
}
